File: src/main/java/mods/eln/gui/ItemStackFilter.java
package mods.eln.gui;

import mods.eln.misc.Utils;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.oredict.OreDictionary;

import java.util.ArrayList;

public class ItemStackFilter {

	int itemId;
	int damageMask;
	int damageValue;

	public ItemStackFilter(Item item, int damageMask, int damageValue) {
		this.itemId = Item.getIdFromItem(item);
		this.damageMask = damageMask;
		this.damageValue = damageValue;
	}

	public ItemStackFilter(Block block, int damageMask, int damageValue) {
		this.itemId = Utils.getItemId(block);
		this.damageMask = damageMask;
		this.damageValue = damageValue;
	}

	public ItemStackFilter(Item item) {
		this.itemId = Item.getIdFromItem(item);
		this.damageMask = 0;
		this.damageValue = 0;
	}

	public ItemStackFilter(Block block) {
		this.itemId = Utils.getItemId(block);
		this.damageMask = 0;
		this.damageValue = 0;
	}

	public static ItemStackFilter[] OreDict(String name) {
		final ArrayList<ItemStack> ores = OreDictionary.getOres(name);
		ItemStackFilter[] filters = new ItemStackFilter[ores.size()];
		for (int i = 0; i < ores.size(); i++) {
			filters[i] = new ItemStackFilter(ores.get(i).getItem(), 0xff, ores.get(i).getItemDamage());
		}
		return filters;
	}

	public boolean tryItemStack(ItemStack itemStack) {// caca1.5.1
		if (Utils.getItemId(itemStack) != itemId)
			return false;
		if ((itemStack.getItemDamage() & damageMask) != damageValue)
			return false;
		return true;
	}
}

File: src/main/java/mods/eln/gui/ItemStackFilter.kt
package mods.eln.gui

import mods.eln.misc.Utils
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraftforge.oredict.OreDictionary

data class ItemFilter(val itemId: Int, val damageMask: Int, val damageValue: Int)

class ItemStackFilter {

    private var whitelist: Array<ItemFilter>

    constructor(item: Item, damageMask: Int = 0, damageValue: Int = 0) {
        whitelist = arrayOf(ItemFilter(
            Item.getIdFromItem(item),
            damageMask,
            damageValue))
    }

    constructor(block: Block, damageMask: Int = 0, damageValue: Int = 0) {
        whitelist = arrayOf(ItemFilter(
            Utils.getItemId(block),
            damageMask,
            damageValue))
    }

    constructor(oreName: String) {
        whitelist = OreDictionary.getOres(oreName).map {
            ItemFilter(Item.getIdFromItem(it.item), 0xff, it.itemDamage)
        }.toTypedArray()
    }

    fun tryItemStack(itemStack: ItemStack): Boolean {
        return whitelist.any {
            Utils.getItemId(itemStack) == it.itemId && (itemStack.itemDamage and it.damageMask) == it.damageValue
        }
    }

    fun add(oreName: String) = apply {
        whitelist = whitelist.plus(ItemStackFilter(oreName).whitelist)
    }

    fun add(item: Item, damageMask: Int = 0, damageValue: Int = 0) = apply {
        whitelist = whitelist.plus(ItemStackFilter(item, damageMask, damageValue).whitelist)
    }

    fun add(block: Block, damageMask: Int = 0, damageValue: Int = 0) = apply {
        whitelist = whitelist.plus(ItemStackFilter(block, damageMask, damageValue).whitelist)
    }

    /**
     * De-duplicates the whitelist. Use if you suspect there may be duplicates.
     */
    fun optimize() {
        whitelist = whitelist.distinct().toTypedArray()
    }
}
